subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,315 @@
+	uint16_t sample_size, total_size;
+	uint16_t conn_sample_size = 0;
+	int i = 0;
+	void *tmp_data;
+
+	__ASSERT(sensor && sensor->dev, "sensor or sensor dev is invalid");
+	sensor_ctx = sensor->dev->data;
+	__ASSERT(sensor_ctx, "sensing sensor context is invalid");
+
+	sample_size = sensor_ctx->register_info->sample_size;
+	for (i = 0; i < sensor->reporter_num; i++) {
+		conn_sample_size += get_reporter_sample_size(sensor, i);
+	}
+
+	/* total memory to be allocated for a sensor according to sensor device tree:
+	 * 1) sample data point to struct sensing_sensor->data_buf
+	 * 2) size of struct sensing_connection* for sensor connection to its reporter
+	 * 3) reporter sample size to be stored in connection data
+	 */
+	total_size = sample_size + sensor->reporter_num * sizeof(*sensor->conns) +
+		     conn_sample_size;
+
+	/* total size for different sensor maybe different, for example:
+	 * there's no reporter for physical sensor, so no connection memory is needed
+	 * reporter num of each virtual sensor may also different, so connection memory is also
+	 * varied, so here malloc is a must for different sensor.
+	 */
+	tmp_data = malloc(total_size);
+	if (!tmp_data) {
+		LOG_ERR("malloc memory for sensing_sensor error");
+		return -ENOMEM;
+	}
+	sensor->sample_size = sample_size;
+	sensor->data_buf = tmp_data;
+	sensor->conns = (struct sensing_connection *)((uint8_t *)sensor->data_buf + sample_size);
+
+	tmp_data = sensor->conns + sensor->reporter_num;
+	for (i = 0; i < sensor->reporter_num; i++) {
+		sensor->conns[i].data = tmp_data;
+		tmp_data = (uint8_t *)tmp_data + get_reporter_sample_size(sensor, i);
+	}
+
+	if (tmp_data != ((uint8_t *)sensor->data_buf + total_size)) {
+		LOG_ERR("sensor memory assign error");
+		return -EINVAL;
+	}
+
+	LOG_INF("create sensor, sensor:%s, min_ri:%d(us)",
+		sensor->dev->name, sensor->info->minimal_interval);
+
+	sensor->interval = 0;
+	sensor->sensitivity_count = sensor_ctx->register_info->sensitivity_count;
+	__ASSERT(sensor->sensitivity_count <= CONFIG_SENSING_MAX_SENSITIVITY_COUNT,
+		 "sensitivity count:%d should not exceed MAX_SENSITIVITY_COUNT",
+		 sensor->sensitivity_count);
+	memset(sensor->sensitivity, 0x00, sizeof(sensor->sensitivity));
+
+	sensor->state = SENSING_SENSOR_STATE_OFFLINE;
+	sys_slist_init(&sensor->client_list);
+
+	sensor_ctx->priv_ptr = sensor;
+
+	return 0;
+}
+
+static int sensing_init(void)
+{
+	struct sensing_context *ctx = &sensing_ctx;
+	struct sensing_sensor *sensor;
+	enum sensing_sensor_state state;
+	int ret = 0;
+	int i = 0;
+
+	LOG_INF("sensing init begin...");
+
+	if (ctx->sensing_initialized) {
+		LOG_INF("sensing is already initialized");
+		return 0;
+	}
+
+	if (ctx->sensor_num == 0) {
+		LOG_WRN("no sensor created by device tree yet");
+		return 0;
+	}
+
+	STRUCT_SECTION_FOREACH(sensing_sensor, tmp_sensor) {
+		ret = pre_init_sensor(tmp_sensor);
+		if (ret) {
+			LOG_ERR("sensing init, create sensor error");
+			return -EINVAL;
+		}
```

Comment:
This path needs to free up any memory allocated for sensors that successfully finished `pre_init_sensor()`.